### PR TITLE
Test ansible-galaxy against pulp/pulp_ansible

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Test ansible-galaxy against pulp/pulp_ansible

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
many

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by jimi-c** [comment]
Needs a rebase.

File: lib/ansible/cli/galaxy.py [hunk 1]
```diff
@@ -447,6 +447,14 @@ def server_config_def(section, key, required):
             auth_url = server_options.pop('auth_url', None)
             token_val = server_options['token'] or NoTokenSentinel
             username = server_options['username']
+            available_api_versions = None
+            v3 = server_options.pop('v3', None)
+            if v3:
+                # This allows a user to explicitly indicate the servers uses the /v3 API
+                # This was added for testing agains pulp_ansible and I'm not sure it has
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]:
```suggestion
                # This was added for testing against pulp_ansible and I'm not sure it has
```

Comment (sivel) [reply lib/ansible/cli/galaxy.py]: Addressed, along with another typo

**Comment by sivel** [comment]
The pulp container is now built using docker. I've pinned it using a digest for now, instead of just using latest.

File: test/lib/ansible_test/_internal/docker_util.py [hunk 2]
```diff
@@ -135,7 +135,7 @@ def docker_get(args, container_id, src, dst):
                     options=['-i'], stdout=dst_fd, capture=True)
 
 
-def docker_run(args, image, options, cmd=None):
+def docker_run(args, image, options, cmd=None, create_only=False):
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/docker_util.py:138]: Add `create_only` to the type hints in the docstring.

File: test/lib/ansible_test/_internal/docker_util.py [hunk 3]
```diff
@@ -160,6 +165,27 @@ def docker_run(args, image, options, cmd=None):
     raise ApplicationError('Failed to run docker image "%s".' % image)
 
 
+def docker_start(args, container_id, options):
+    """
+    :type args: EnvironmentConfig
+    :type container_id: str
+    :type options: list[str] | None
+    :rtype: str | None, str | None
+    """
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/docker_util.py:172]: New functions should use PEP 484 type hints (comment based) and provide a description in the docstring instead.

File: test/lib/ansible_test/_internal/docker_util.py [hunk 4]
```diff
@@ -275,7 +301,7 @@ def docker_version(args):
     return json.loads(stdout)
 
 
-def docker_command(args, cmd, capture=False, stdin=None, stdout=None, always=False):
+def docker_command(args, cmd, capture=False, stdin=None, stdout=None, always=False, data=None):
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/docker_util.py:302]: Add `data` to the type hints in the docstring.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/lib/ansible_test/_internal/docker_util.py (hunk 4 above)
Comment (sivel) [reply test/lib/ansible_test/_internal/docker_util.py:302]: Done.

File: test/lib/ansible_test/_internal/docker_util.py (hunk 3 above)
Comment (sivel) [reply test/lib/ansible_test/_internal/docker_util.py:172]: Done, I think :)

File: test/lib/ansible_test/_internal/docker_util.py (hunk 2 above)
Comment (sivel) [reply test/lib/ansible_test/_internal/docker_util.py:138]: done.

**Review by mattclay — APPROVED** [review]
The changes look good overall, but it looks like there are issues with the tests.

One of the test failures shows the error:

```
Error when publishing collection to cmd_arg (http://fallaxy-stub:8080/api/) (HTTP Code: 500, Message: INTERNAL SERVER ERROR Code: Unknown)
```

**Comment by sivel** [comment]
> The changes look good overall, but it looks like there are issues with the tests.
> 
> One of the test failures shows the error:
> 
> ```
> Error when publishing collection to cmd_arg (http://fallaxy-stub:8080/api/) (HTTP Code: 500, Message: INTERNAL SERVER ERROR Code: Unknown)
> ```

Yay, another random transient error in fallaxy that I have no idea what caused it.

**Comment by sivel** [comment]
That error is not anything new with this PR though.

**Comment by sivel** [comment]
I've restarted tests, and everything is passing, if fallaxy continues to be a problem, I can troubleshoot it more at a later date.

**Pull request merged**